signac/core/collection_api.py

```diff
@@ -0,0 +1,491 @@
+        else:
+            self.__delitem__(key)
+
+
+class SyncedList(SyncedCollection, MutableSequence):
+
+    def __init__(self, **kwargs):
+        data = kwargs.pop('data', None)
+        super().__init__(**kwargs)
+        if data is None:
+            self._data = []
+        else:
+            self._data = [self.from_base(data=value, parent=self) for value in data]
+
+    def to_base(self):
+        converted = list()
+        for value in self._data:
+            if isinstance(value, SyncedCollection):
+                converted.append(value.to_base())
+            else:
+                converted.append(value)
+        return converted
+
+    def reset(self, data=None):
+        if data is None:
+            data = []
+        if isinstance(data, Sequence) and not isinstance(data, str):
+            backup = copy(self._data)
+            try:
+                with self._suspend_sync():
+                    for i in range(min(len(self), len(data))):
+                        if data[i] == self._data[i]:
+                            continue
+                        try:
+                            self._data[i].reset(data[i])
+                            continue
+                        except (ValueError, AttributeError):
+                            pass
+                        self._data[i] = self.from_base(data=data[i], parent=self)
+                    if len(self._data) > len(data):
+                        self._data = self._data[:len(data)]
+                    else:
+                        self.extend(data[len(self):])
+                self.sync()
+            except BaseException:  # rollback
+                self._data = backup
+                raise
+        else:
+            raise ValueError("The data must be a non-string sequence or None.")
+
+    def __delitem__(self, item):
+        self.load()
+        with self._safe_sync():
+            del self._data[item]
+            self.sync()
+
+    def __setitem__(self, key, value):
+        self.load()
+        with self._safe_sync():
+            with self._suspend_sync():
+                self._data[key] = self.from_base(data=value, parent=self)
+            self.sync()
+
+    def __getitem__(self, key):
+        self.load()
+        return self._data[key]
+
+    def __iter__(self):
+        self.load()
+        return iter(self._data)
+
+    def __len__(self):
+        self.load()
+        return len(self._data)
+
+    def __reversed__(self):
+        self.load()
+        return reversed(self._data)
+
+    def __iadd__(self, iterable):
+        self.load()
+        with self._safe_sync():
+            self._data += [self.from_base(data=value, parent=self) for value in iterable]
+            self.sync()
+
+    def __call__(self):
+        self.load()
+        return self.to_base()
+
+    def __eq__(self, other):
+        if isinstance(other, type(self)):
+            return self() == other()
+        else:
+            return self() == other
```

Comment:
No need to address this now, but file away for later: this could be a significant performance bottleneck since it requires a conversion to the base type for each equality check.

Reply:
Yes, I believe this is copied verbatim.

Reply:
This is copied verbatim. Marking resolved.